RecoEcal/EgammaCoreTools/interface/EcalClusterTools.h

```diff
@@ -332,61 +326,36 @@ float EcalClusterToolsT<noZS>::recHitEnergy(DetId id, const EcalRecHitCollection
 //    -2 |_|_|_|_|_|
 //      -2 -1 0 1 2 ix
 template<bool noZS>
-float EcalClusterToolsT<noZS>::matrixEnergy( const reco::BasicCluster &cluster, const EcalRecHitCollection *recHits, const CaloTopology* topology, DetId id, int ixMin, int ixMax, int iyMin, int iyMax )
+float EcalClusterToolsT<noZS>::matrixEnergy( const reco::BasicCluster &cluster, const EcalRecHitCollection *recHits, const CaloTopology* topology, DetId id, CaloRectangle rectangle )
 {
-  //take into account fractions
-    // fast version
-    CaloNavigator<DetId> cursor = CaloNavigator<DetId>( id, topology->getSubdetectorTopology( id ) );
     float energy = 0;
-    const std::vector< std::pair<DetId, float> >& v_id = cluster.hitsAndFractions();
-    for ( int i = ixMin; i <= ixMax; ++i ) {
-        for ( int j = iyMin; j <= iyMax; ++j ) {
-	  cursor.home();
-	  cursor.offsetBy( i, j );
-	  float frac=getFraction(v_id,*cursor);
-	  energy += recHitEnergy( *cursor, recHits )*frac;
-        }
+    auto const& vId = cluster.hitsAndFractions();
+
+    for (auto const& detId : rectangle(id, *topology)) {
+        energy += recHitEnergy( detId, recHits ) * getFraction(vId, detId);
     }
-    // slow elegant version
```

Comment:
I was a bit surprised by this comment here: why was that solution so slow? Is looping over the DetIDs twice and keeping a vector of them for the second loop so expensive? That statement prompted me to do this rectangle range solution where you only loop over the detIDs once and can use then for what you like in this loop. I hope it combines the speed of the previous solution with the "elegance" of the commented-out solution.

Reply:
So this solution is slow because a cluster can own many many hits due to how PF clustering works. As you know, PF clustering shares the energy of a crystal between local energy maximums. It does this fairly globally and (initially at least) did not clean well. So you would end up with half the ECAL assigned as a "rec-hit" of the cluster but with fraction of 1E-8 or similar. This was later cleaned up to be >1E-4/E-5 (and may have been further changed) but at the time of writing you were running over a huge number of hits, hence the "slow" comment. Its not clear to me if this was actually measured to be slow or just thought to be from the above argument, I think it was actually slow from dimly remembered conversations. 

One thing which I want to achieve in the shutdown is to review the minimum fraction (or fraction*energy) for a hit to be part of a PFCluster to further reduce this problem. Although due to the raised minimum fraction this may not be a problem anymore.

Reply:
That is interesting, thanks for explaining! But I was talking about a different thing. You explained why `recHitEnergy()` can be slow, right? I was more talking about the fact that the comments labels this as __slow__ (if you "unwrap" the functions and remove the `frac` related stuff):
```
CaloNavigator<DetId> cursor = CaloNavigator<DetId>( id, topology->getSubdetectorTopology( id ) );
std::vector<DetId> v;
 for ( int i = ixMin; i <= ixMax; ++i ) {
    for ( int j = iyMin; j <= iyMax; ++j ) {
         cursor.home();
         cursor.offsetBy( i, j );
         if ( *cursor != DetId(0) ) v.push_back( *cursor );
    }
}

float energy = 0;
for ( std::vector<DetId>::const_iterator it = v.begin(); it != v.end(); ++it ) {
        energy += recHitEnergy( *it, recHits );
}
```
and this as ___fast__:
```
float energy = 0;
CaloNavigator<DetId> cursor = CaloNavigator<DetId>( id, topology->getSubdetectorTopology( id ) );
std::vector<DetId> v;
 for ( int i = ixMin; i <= ixMax; ++i ) {
    for ( int j = iyMin; j <= iyMax; ++j ) {
         cursor.home();
         cursor.offsetBy( i, j );
         energy += recHitEnergy( *cursor, recHits )
    }
}
```

In the "slow" solution, the DetIds in the rectangle are precomputed and stored, and in the "fast" solution the energies are already added in the loop where the DetHits are obtained (and the DetHits are not stored). If one believes that this really makes a difference, it made sense to implement this CaloRectangleRange I thought.